Edit/register

yangjinhwa opened the pull request:
1. fixed Promise then function to async await
2. throw error (message or `ErrorModal`)
3. arrange the code according to the lint

yangjinhwa committed:
refect promise function to async/await function and throw error

yangjinhwa committed:
node version update

HyunmoAhn commented on package.json:
```diff
@@ -60,7 +60,7 @@
     "eslint-plugin-jsx-a11y": "^6.3.1",
     "eslint-plugin-prettier": "^3.1.4",
-    "eslint-plugin-react": "^7.20.6",
+    "eslint-plugin-react": "^7.2sud0.6",
```
?!

yangjinhwa replied:
sorry....

HyunmoAhn commented on src/component/modal/ErrorModal.jsx:
```diff
@@ -0,0 +1,29 @@
+  return (
+    <>
+      <div className="ModalOverlay">
+        <div className="ModalWrapper ModalAni_pop" onClick={(e) => e.stopPropagation()}>
```
`stopPropagation`은 왜 필요한건가요?

yangjinhwa replied:
필요없어보입니다...!

HyunmoAhn commented on src/component/modal/ErrorModal.jsx:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  return (
+    <>
```
fragment 지워도 될 것 같아요

HyunmoAhn commented on src/component/register/signIn/index.jsx:
```diff
@@ -44,22 +45,21 @@ function SignIn() {
+  }
+
+  function handleEnterKeyPress(e) {
+    if (e.key === 'Enter') {
```
keyCode를 사용하는게 좋을 것 같아요. string 비교보다는 code 비교가 좋다고 생각해요.

그리고 code나 string은 상수로 만들어서 `const KEY_CODE = { ENTER: xx }` 이렇게 사용하는게 좋아요.

yangjinhwa replied:
[MDN의 e.keyCode 문서](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) 를 보면 keyCode는 더이상 추천하지 않는다고 합니다. 또다른 방법인 [e.witch](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which) 방법 역시 추천되지 않았습니다. e.key와 e.keyCode를 비교분석하는 글들은 꽤 오래되었는데 대부분 keyCode를 지양했습니다...

HyunmoAhn replied:
이건 몰랐네요. 그럼 `Enter`를 상수로 만들어서 비교하면 좋을 것 같아요.

HyunmoAhn commented on src/component/register/signIn/index.jsx:
```diff
@@ -44,22 +45,21 @@ function SignIn() {
+
+  function handleEnterKeyPress(e) {
+    if (e.key === 'Enter') {
+      requestLogin().then();
```
`then()`은 필요 없을 것 같아요

yangjinhwa replied:
requestLogin() 이 Promise여서 then이 필요한게 아닐까요!? 일단 then()없이는 에러가 납니다

HyunmoAhn replied:
어떤 에러가 발생하나요? 안나는게 정상인데

yangjinhwa replied:
![image](https://user-images.githubusercontent.com/29908256/94349290-1568c880-007e-11eb-8766-dc376e5c7710.png)
Promise여서 then으로 실행해주어야한다는것 아닐까요?

HyunmoAhn replied:
이건 오류가 아니라, IDE에서 해주는 경고 같은 건데요.

`requestLogin`함수가 Promise로 리턴되지만 requestLogin을 await로 호출한 것도 아니며, 뒤에 then이 있는 것도 아니니 비동기함수를 비동기로 처리하고 있지 않으니 `requestLogin`의 Promise가 무의미하다는 의미인데 의도에 따라서는 IDE의 경고를 무시해도 된다고 생각해요. 

경고를 지우고 싶으면 await로 호출하면 되겠죠.

HyunmoAhn commented on src/component/register/signIn/index.jsx:
```diff
@@ -77,9 +77,11 @@ function SignIn() {
-          placeholder="영문/숫자 혼용 8자 이상"
+          placeholder="비밀번호(영문/숫자/특수문자 혼합 8자 이상)"
+          maxLength={30}
           onChange={validatePassword}
```
여기 변경사항에는 빠져있는데, utils 파일에 validation관련 함수를 모아서 true/false를 판단하게 유틸함수로 빼서 사용하는게 좋을 것 같아요.

유효성 검사 같은 경우는 여러 화면에서 사용할 수 있어서요.

yangjinhwa replied:
음 요것은 나중에 특정 유효성이 반복되는 시점(예를들면 설정 등)에서 리펙토링 해도 될까요?

HyunmoAhn replied:
지금도 `signin/index.jsx`, `paswordInput.jsx`, `emailInput` 여기 3군데에서 email, password 유효성검사가 각각 2번씩 반복되는데 나중에 하고 싶은 이유가 있나요?

yangjinhwa replied:
아아 그러네요! 한 번만 쓰이는줄 착각했습니다ㅠ 따로 빼도록 할게요~!

HyunmoAhn commented on src/component/register/signIn/index.jsx:
```diff
@@ -77,9 +77,11 @@ function SignIn() {
+          maxLength={30}
           onChange={validatePassword}
           value={passwordValue}
+          onKeyPress={handleEnterKeyPress}
```
비밀번호 입력할때만 enter로 로그인을 진행하는건가요? id에도 필요하지 않나요?

yangjinhwa replied:
유효성이 모두 통과되었을때만 로그인 또는 회원가입 버튼이 활성화 되므로, 유효성이 통과되지 않은 input(email, name등)에서 enter로 접근할 일이 적다고 생각했습니다. 그래서 유저 사용성중 비교적 마지막으로 유효성 검증을 마칠 확률이 가장 큰 password 에만 enter를 걸었습니다.

HyunmoAhn commented on src/component/register/signUp/emailInput.jsx:
```diff
@@ -35,29 +35,34 @@ function SignUpEmailInput({ setEmailAuthId, emailAuthId, emailValue, setEmailVal
-    } catch (e) {
-      console.log(e);
+    const { existence } = await fetchEmailExistence(emailValue);
+    if (!existence) {
```
`setIsEmailExistence`는 조건문 밖에서 

```
setIsEmailExistence(existence);
```
이렇게 설정하는게 좋을 것 같아요.

HyunmoAhn commented on src/component/register/signUp/emailInput.jsx:
```diff
@@ -35,29 +35,34 @@ function SignUpEmailInput({ setEmailAuthId, emailAuthId, emailValue, setEmailVal
+
+  function handleEmailAuthEnterKeyPress(e) {
+    if (e.key === 'Enter') {
+      handleEmailAuth().then();
```
여기도 key를 keyCode로 사용해주시고, code는 상수화 해주세요.

then()은 필요 없으니까 빼주시구요.

HyunmoAhn commented on src/component/register/signUp/index.jsx:
```diff
@@ -22,18 +23,32 @@ function SignUp() {
+
+  function handleEnterKeyPress(e) {
+    if (e.key === 'Enter') {
+      submitUser().then();
```
여기도 마찬가지로 key -> code로 변경, 상수화, then() 삭제 부탁드려요

yangjinhwa replied:
enter 상수화 했습니다!

HyunmoAhn commented on src/component/registerEntry/index.jsx:
```diff
@@ -11,10 +12,19 @@ function RegisterEntry() {
 
+  const [userName, setUserName] = useState('');
+  useEffect(() => {
+    getUserName().then();
```
then()이 필요없으면 적을 필요 없어요.

HyunmoAhn commented on src/constants/modalData.js:
```diff
@@ -8,3 +8,14 @@ export const MODAL_DATA = {
+  LOGIN_ERROR: {
+    title: '로그인 안내',
+    description: `아직 가입하지 않았거나 비밀번호를 잘못 입력했어요.
+    확인하고 다시 입력해 주세요.`,
```
내부에 변수가 없으면 string은 `대신 '를 사용해주세요.

HyunmoAhn commented on src/store/api/registerApi.js:
```diff
@@ -1,13 +1,27 @@
+
+  const groupData = await fetchGroupMe();
+  const { count: userGroupCount } = groupData;
+  setGroupCount(userGroupCount);
```
GroupCount는 어떤 정보인가요?

token값을 cookie에 저장하는건 알겠는데, groupCount도 cookie에 저장해야하나 싶어서요.

yangjinhwa replied:
groupCount === 0 일 시 registerEntry 페이지로 던져야 합니다. (serviceMain진입 전 그룹만들기 또는 참여하기가 필수임)
groupCount !== 0 일 시 group정보를 들고와 serviceMain에 보여줍니다.

이 때 그룹을 만드는 각 페이지 (registerEntry, makeGroup)에서도 token 및 groupCount를 들고있어야 페이지로의 접근이 가능하기 때문에 각 페이지로 넘어갈때마다 groupCount값을 가져오기 위해 통신하기 보다는 localStorage에 저장하는것이 더 효율적이라고 생각했습니다...!

HyunmoAhn replied:
`groupCount`는 사용자가 가지고 있는 그룹의 개수이네요.

그렇다면, 더더욱 client에 저장해놓는 것이 아니라 서버에서 불러와야하는게 맞는 것 같아요.

사용자가 하나의 client만 사용할 것이라는 생각을 하면 안되고, 여러 환경에서 로그인을 할 수 있기 때문에, token과 같은 인증 정보는 서버에서 만료를 컨트롤 할 수 있기 때문에 문제가 될 것 같진 않은데요.

사용자가 A, B 브라우저에서 로그인을 하고 B 브라우저에서 그룹을 만들었다면, B 브라우저의 카운트만 올라가고 A 브라우저에서는 groupCount가 0이 되니까 예상하지 못하는 이슈가 발생할 가능성이 높아요.

따라서 groupCount는 cookie형태로 저장하는게 아니라 변수 형태로 저장하는게 맞다고 생각하고, 로그인시마다 혹은 페이지 접근시마다 서버에서 데이터를 받는게 맞을 것 같아요.

물론, 페이지 내부를 왔다갔다할때마다 서버로부터 정보를 가져올 수는 없으니 한곳에서 데이터를 관리할 방법을 생각해야 할 거구요.

yangjinhwa replied:
아하...다른 client에서의 접근은 고려 못했던 사항인것 같습니다..!
그렇다면 역시 접근 권한을 가지는 페이지마다 요청이 발생해야 할 것 같은데, 맨 마지막줄에 한곳에서 데이터를 관리한다는게 어떤 의미인지 궁금합니다...!

HyunmoAhn replied:
저희 프로젝트는 SPA니까 유저에 대한 정보는 처음 로그인시에 받아 온 뒤, redux의 store와 비슷한 역할을 하는 저장공간에 저장한 후 내부 페이지를 이동할때 그 저장공간에서 유저 정보를 꺼내쓰면 된다고 생각해요.(매번 요청하는데는 페이지마다 로딩이 필요하므로)

만약에 여러 브라우저에서 세션이 유지되는 것이 정책이고, 유저 정보가 실시간으로 반영되어야하는 민감한 정보라면 페이지 접근시마다 정보를 서버로 요청해서 가져와야겠지만, 일반적으로는 로그인시 정보를 받아온 후 저장해 놓은 정보를 사용하고 서버에서 던져주는 에러메세지(빈도가 그렇게 많지 않음)를 가지고 처리하는 편인 것 같아요.

즉, 한번 서버로 요청해서 데이터를 받아 온 뒤, 그걸 어디다 저장해놓고 계속 꺼내쓰는거죠.

yangjinhwa replied:
여기가....redux 도입 시점 일까요....?

HyunmoAhn replied:
redux가 글로벌하게 상태를 관리해주는 도구는 맞지만, 글로벌하게 상태 관리를 할 방법이 redux만 있는건 아니에요.

`app.js`에서 state를 만들어두고 props drilling을 통해서 내부 컴포넌트로 넣어줘서 전역에서 사용할 수 도 있고,
react의 context기능을 이용해서 전역에서 사용 할 수도 있구요.
utils위치에 global한 데이터 저장소를 두고 import를 해서 사용 할 수도 있구요.
지금처럼 cookie에 정보를 저장해서 사용하는 것도 일종의 global하게 상태 관리를 하는거구요. (이거는 위에 말했다시피 문제가 있는거구요.)

그 방법에 대해서 고민을 해보야아 할거에요.

yangjinhwa replied:
위의 말씀을 살펴보고 loading구현에 대해서 생각해보고 있는데 분기점마다 결국 loading으로 감싸는 컴포넌트 위계에 따라서 drilling이 심할수도 있을것 같단 생각이 듭니다ㅠㅠ현재단계에서는 loading을 최대한 크게 쪼개는게..나을까요?

HyunmoAhn replied:
loading에 관련해서는 현재 login과 연관이 있다기 보단 프로젝트 전반적으로 영향이 있을 것 같다는 생각이 들며, 방향에 따라서 프로젝트의 구조를 크게 변경해야 할 수 있습니다.

global store 생성에 대해서도 동일한 의견이라고 생각합니다.

따라서, 현재 PR에서 해당 문제를 해결하고 넘어가기 보단, [issue](https://github.com/our-library/library-web/issues)에 global store생성에 대한 이슈, loading을 어떻게 처리할지에 대한 이슈 총 2개를 생성해서 필요한 이유 및 적용방법에 대한 후보, 적용 방법들에 대한 장, 단점 혹은 고민하고 있는 이유등 정리해보는건 어떨까요?

현재 PR에 리뷰가 쌓여가는데 로그인 기능에 대해서 정리된 부분은 마무리하고 새롭게 PR을 여는게 나을 것 같아요.

yangjinhwa replied:
확인했습니다. 해당 PR 업데이트 후 새로 열겠습니다!

HyunmoAhn committed:
Merge branch 'master' into edit/register

HyunmoAhn commented:
맨처음 `/register`로 갈때 signIn, signUp 둘 다 아니라서 `NavLink`가 활성화가 안되는데, 이거 버그죠? 확인해주세요.

그리고, Link에 쓰이는, Route에 쓰이는 path는 상수로 변경해주세요.

HyunmoAhn commented on src/component/register/signUp/emailInput.jsx:
```diff
@@ -35,29 +35,34 @@ function SignUpEmailInput({ setEmailAuthId, emailAuthId, emailValue, setEmailVal
-      })
-      .catch((e) => console.log(e));
+  async function completeEmailAuth() {
+    await fetchEmailAuthorize(emailAuthId, emailAuthKey);
```
이거 요청 response가 json 형태가 아닌 것 같은데 확인해주세요.

서버 개발자가 API를 뚫어놓고 직접 테스트는 client에서 진행하기때문에 서버에서 response가 제대로 오는지도 client 개발시에 확인해줘야하는 부분이라고 생각해요.

yangjinhwa replied:
fetchEmailAuthorize는  response값이 따로 없는 post요청 맞습니다!

HyunmoAhn replied:
response를 받지 않더라도 정상일때 json형태로 응답을 받는게 맞다고 생각해요.

yangjinhwa replied:
응답이라 함은 status code(200)을 말하시는 걸까요..?

HyunmoAhn replied:
![image](https://user-images.githubusercontent.com/23732795/94355111-1a4c6d00-00bc-11eb-9b35-92a1fe0eec32.png)

위의 response형식이 제가 사용하고 있는 API의 예시인데요. 성공했을때 정보를 받을 필요가 없는데 content에 빈 정보를 내려주고 있거든요. 특정 API는 json형태로 받고, 특정 API는 json형태로 받지않는 것 보다는 하나로 통일을 하는게 좋다고 생각했어요. 크게 문제가 아니라고 생각하면 넘어가시죠.

yangjinhwa replied:
오호 요것은 참고해서 나중에 리나랑 얘기해보겠습니다!

HyunmoAhn commented:
로그인 세션 유지 스펙이 어떻게 되나요?

지금 로그인 하고 뒤로가기를 누르거나 새로고침을 통해서 로그인페이지로 이동이 가능해서요.

HyunmoAhn commented on src/component/registerEntry/index.jsx:
```diff
@@ -26,19 +36,23 @@ function RegisterEntry() {
-        <button type="button" onClick={() => history.replace('/makeGroup')} className="Btn-md Btn-primary">
+        <button
+          type="button"
+          onClick={() => history.replace('/makeGroup')}
```
url은 대소문자 구분을 안하는 것으로 알고 있어요.
그래서 url에서는 camelCase보다 kebab-case를 쓰는게 일반적이에요

지금 github url만 봐도 이렇게 되있어요.
![image](https://user-images.githubusercontent.com/23732795/94345135-ce1f0f80-005e-11eb-985e-a29c94545c39.png)

yangjinhwa replied:
확인했습니다! 수정할게욥

HyunmoAhn commented:
로딩... 은 아직 안넣는거죠? 새로고침하면 이름이나 정보들이 비어있다가 보이는게, 로딩이 필요 할 것 같아서요.

HyunmoAhn commented on src/component/registerEntry/jobNameInput.jsx:
```diff
@@ -48,14 +52,14 @@ function JobNameInput({ setJobKey, setIsValidJopNameValue }) {
           </ul>
         </div>
       )}
+      {isJobNameKeyValid && <JobNameNoKeyErrorMessage />}
```
jobName은 select가 아니라 input인가요? 어떤게 유효한 값인지 알기 어려울것 같은데요?

yangjinhwa replied:
넵 input입니다. 미리 작성된 jobList: Object 에서 jobKey값과 jobName값을 매칭합니다.
jobKey와 매칭되지 않는 jobName은 api로 발송이 불가하고 요청이 jobKey:number 로만 요청 가능하기 때문에 
해당 로직은 jobname이 매칭되는 jobKey가 없는 경우 요청이 반려된다는 로직이었습니다...! 변수명이 알기 어려운걸까요?

HyunmoAhn replied:
사용자는 어떤 job list가 있는 줄 알고 어떤 값을 넣나요? 

로직과 관련된 코멘트가 아니라, 사용성과 관련된 질문이었어요.

yangjinhwa replied:
현재는 검색 시 자동완성의 느낌으로 input value를 포함하는 jobName 리스트가 뜨고, 클릭하면 input에 반영되는 식입니다.

HyunmoAhn replied:
아... 한글 자음만 넣다보니까 아래 리스트가 안 보였네요. 기획이 그러면 넘어갈게요. 👍 

올바르지 않은 코드나, job을 넣었을때 서버에서 에러를 던져주는데 그에 대한 처리는 아직 미정인가요?

yangjinhwa replied:
![image](https://user-images.githubusercontent.com/29908256/94371346-da739d00-0130-11eb-89a0-455865b41c02.png)
코드값은 빈값외에는 넘어갈 일이 없어보여서.. 일단은 커스텀되어 jobKey 빈값처리하는 구문은 위에 있습니다!

HyunmoAhn replied:
![image](https://user-images.githubusercontent.com/23732795/94373284-6260a380-013f-11eb-882a-8ad736a51b40.png)

제가 이야기하하는건, 잘못된 요청을 보냈을때 사용자에게 보여주는 에러처리를 말한거였습니다.
위에 스크린샷을 보면 400 bad request가 왔는데 화면에는 아무런 표시를 하고 있지 않아요.

yangjinhwa replied:
아하 확인했습니다. 에러처리할게욥!

yangjinhwa replied:
유효성 처리에 대한 기획 검토 후 수정될 예정입니다!

HyunmoAhn commented on src/constants/modalData.js:
```diff
@@ -8,3 +8,14 @@ export const MODAL_DATA = {
+    isCloseBtn: false,
+    confirmBtnName: '다시 입력할게요.',
+  },
+};
```
constants라는 폴더명이 상수 데이터들을 모아두는 네이밍이니까 modalData.js보다는 modal.js 가 어때요?

yangjinhwa replied:
좋은의견입니다. 반영하겠습니다

yangjinhwa committed:
replace route path to constants

yangjinhwa committed:
conflict resolution

yangjinhwa committed:
fixed conflict

yangjinhwa committed:
Merge branch 'master' into edit/register

yangjinhwa committed:
change route path to constants

yangjinhwa committed:
change route path to constants and arrange async functions

HyunmoAhn commented on src/component/registerEntry/index.jsx:
```diff
@@ -23,7 +26,7 @@ function RegisterEntry() {
   async function handleJoinGroup() {
     await joinGroupRequest(inviteCode, jobKey);
-    history.push('/service');
+    history.push(SERVICE);
```
취향차이긴한데요. 
저는 ROUTE_PATH를 굳이 destructuring하지 않고 `ROUTE_PATH.SERVICE` 이렇게 해서 이게 path와 관련된 정보다 라고 명시하는 편이에요. 참고참고

HyunmoAhn commented on src/routes/protectedRoute.jsx:
```diff
@@ -18,18 +18,24 @@ export function ProtectedRoute({ children, ...rest }) {
-  const userGroupCount = getGroupCount();
   const { SERVICE } = ROUTE_PATH;
+  const [groupCount, setGroupCount] = useState(0);
+  getUserGroupCount().then((result) => setGroupCount(result));
```
`RegisterProtectedPage`가 업데이트 될 때마다 API요청을 보낼텐데 괜찮나요....?

yangjinhwa replied:
[위의 말씀](https://github.com/our-library/library-web/pull/10#discussion_r495522498) 처럼 전역으로 데이터가 관리되는 store가 없는 경우에는 어쩔수없지 않나 싶습니다...

HyunmoAhn replied:
1. 없으면 만들면 됩니다. 
2. 이건 데이터를 global하게 관리하는 문제가 아니라 react의 lifecycle과 관련된 이슈입니다.

2번에 대해서 좀 더 설명을 하면, react의 함수형 컴포넌트에는 useEffect라는 라이프사이클 훅이 존재하고, 이를 이용해서 모든 업데이트마다 요청이 발생하는 것을 방지하게끔 개발을 해야합니다.

업데이트마다 서버로 요청보내는 코드를 삭제할 수 없다고 말씀하시는 것은 `react useEffect`에 대한 공부가 필요하다고 생각됩니다.

해결책을 말씀드리면,
`getUserGroupCount`는 서버로 요청을 보내서 `groupCount` 결과를 가져옵니다. 
그렇다면 `groupCount`값이 변경되었다고 생각하는 타이밍에 서버로 요청을 보내면 되고, useEffect의 dependency에 해당 변수를 넣어주어서 `groupCount`값이 변경되었을 때에만 서버로 요청을 보내는게 맞는겁니다.
그 타이밍이 결코 `protectedRoute` 컴포넌트가 업데이트 되는 모든 시점은 아닐거라고 단언할 수 있습니다.

yangjinhwa replied:
음...제가 잘못 이해하고 드린 답변인지 모르겠는데 저는 기존에 `RegisterProtectedPage`를 다른 client에서 접근했을때의 처리에 대한 맥락에서 생각해 groupCount를 로그인 시 한번만 불러내고 저장해 전역에서 관리하는 방면에 대한(1번) 말씀이라고 생각했습니다. 그래서 전역으로 관리하는 groupCount에 대한 의문을 제기한 것이구요.
2번의 코멘트를 보니 질문주신 바가 `RegisterProtectedPage`의 성능개선에 가깝다는 생각이 드네요. 뷰 랜더링에 대한 비효율에 관한 질문이셨다면 개선하겠습니다!

HyunmoAhn replied:
맨 처음 답변이 이해가 됬습니다. 말씀하신대로 각 페이지별로 groupCount를 불러오는것도 문제가 되고, 그래서 전역으로 관리되는 store가 필요한 것인데요. 제가 코멘트를 단 이유는 조금 다른 이유입니다.

이해하신 것 처럼 현재 스레드의 주제는 `protectedRoute` 컴포넌트 자체의 라이프 사이클때문에 문제가 생길 것 같다는 것이므로 그쪽 수정해서 commit 주시면 확인해보겠습니다 😄

HyunmoAhn commented on src/component/actionBar/index.jsx:
```diff
@@ -26,33 +28,29 @@ function ActionBar() {
-      setUserOverdueRentals(overdueRentals);
-      setGroupInvitationKey(invitationKey);
-    });
+    getUserInfo().then();
```
여기도 then 불필요 한 것 같아요.

HyunmoAhn commented:
@yangjinhwa 

이렇게 접힌 부분에도 코멘트가 있으니 여기도 확인해주세요~ 혹시나 모르실까봐!
![image](https://user-images.githubusercontent.com/23732795/94355302-0f92d780-00be-11eb-9318-260fb7fd9391.png)

yangjinhwa committed:
separate validate email & password function

yangjinhwa committed:
fixed keyCode string to constants
improve invitation code error from error message to modal
add jobname search filter condition toLowerCase

yangjinhwa committed:
improve register-protected-page's performance with useEffect

yangjinhwa committed:
create loading component (just draw not a function)

yangjinhwa committed:
fixed lint

HyunmoAhn commented on src/routes/protectedRoute.jsx:
```diff
@@ -17,17 +18,26 @@ export function ProtectedRoute({ children, ...rest }) {
+  const [groupCount, setGroupCount] = useState(0);
+  useEffect(() => {
+    getUserGroupCount().then((result) => setGroupCount(result));
+  }, [groupCount]);
```
이쪽 dependency 삭제 해야할 것 같은데 어떻게 생각하세요?

yangjinhwa replied:
아아 맞는것 같습니다!

HyunmoAhn reviewed (dismissed):
[이쪽 코멘트](https://github.com/our-library/library-web/pull/10/files#r497348246)만 반영하고 머지하시면 될 것 같아요.

큰 피쳐 고생 많이 하셨습니다! 👍

yangjinhwa committed:
remove protected-route useEffect dependency

yangjinhwa committed:
Update jobName list and Add key-down events when the search result is exist

yangjinhwa committed:
fixed lint

HyunmoAhn commented on src/constants/jobLists.js:
```diff
@@ -1,43 +1,23 @@
 export const jobLists = [
+  {
+    jobKey: 0o1,
```
`0o1`?

yangjinhwa replied:
001로 하니까 아래와같은 에러가 발생해서ㅠ 추천대로 고쳤습니다 자바스크립트 제너럴 에러라고 뜨네요

![image](https://user-images.githubusercontent.com/29908256/94698570-69242c00-0374-11eb-89a8-216d42616650.png)

HyunmoAhn replied:
jobKey가 string이면 '001'이 되어야할거고, 숫자면 그냥 1이면 충분할 것 같습니다.

0o1이 뭘 의미하는지 모르겠지만요.

yangjinhwa replied:
number로서 001은 존재할 수 없나봅니다 1로 대체하도록 할게요

HyunmoAhn commented on src/utils/useKeyPress.jsx:
```diff
@@ -9,19 +9,19 @@ export const useKeyPress = function(targetKey) {
-  useEffect(() => {
-    window.addEventListener("keydown", downHandler);
-    window.addEventListener("keyup", upHandler);
+  React.useEffect(() => {
```
요거는 왜 `React.useEffect`로 사용했나요?

yangjinhwa replied:
001로 하니까 아래와같은 에러가 발생해서ㅠ 추천대로 고쳤습니다 자바스크립트 제너럴 에러라고 뜨네요

![image](https://user-images.githubusercontent.com/29908256/94698180-016de100-0374-11eb-9419-68cdd2c0e800.png)

yangjinhwa replied:
앗 useEffect만 쓰는걸로 바꿨습니다. custom hook예시보면서 하다가 마지막에 고쳤었는데 반영이 안되었네요

HyunmoAhn commented on src/utils/useKeyPress.jsx:
```diff
@@ -0,0 +1,29 @@
+      window.removeEventListener("keydown", downHandler);
+      window.removeEventListener("keyup", upHandler);
+    };
+  });
```
customHook을 이렇게 사용하는건 좋은데요.

keydown이랑 keyup은 동시에 일어나지 않을 수 있어요.
예를들어서 `아래 화살표`를 누른 다음에 브라우저가 아닌 다른 창을 클릭하면 keyup을 해도 브라우저에서 이벤트가 발생을 하지 않아요.

그러면 `keyPressed`가 true인 상태가 유지되기 때문에 다음 `아래 화살표`를 누르면 true -> true가 되기 때문에 동작을 하지 않는거죠.

다른 로직으로 변경해야 할 것 같아요.

yangjinhwa replied:
음 말씀하신 해당 에러가 저는 발생하지 않는 것 같은데 브라우저가 아닌 다른 창을 누른다는게 정확히 어떤 상황일까요?

yangjinhwa committed:
replace useEffect

yangjinhwa committed:
improve job list name search keypress
